### PR TITLE
gh-84753: Make inspect.iscoroutinefunction() work with AsyncMock

### PR DESCRIPTION
**Pull request by sileht**
The inspect version was not working with `unittest.mock.AsyncMock`.

This change moves the `asyncio` fix for `AsyncMock` in inspect module and
make `asyncio.iscoroutinefunction` an alias of `inspect.iscoroutinefunction`.

* Issue: gh-84753

### CODE REVIEW

**Comment by sileht** [comment]
@vstinner Can you take a look?

**Review by jd — APPROVED** [review]

**Comment by graingert** [comment]
this shouldn't be needed AsyncMock already sets https://github.com/python/cpython/blob/5fcfdd87c9b5066a581d3ccb4b2fede938f343ec/Lib/unittest/mock.py#L2175-L2178

File: Lib/inspect.py [hunk 1]
```diff
@@ -406,12 +406,15 @@ def isgeneratorfunction(obj):
     See help(isfunction) for a list of attributes."""
     return _has_code_flag(obj, CO_GENERATOR)
 
+# A marker for iscoroutinefunction.
+_is_coroutine = object()
```
Comment (cjw296) [inline Lib/inspect.py]: Why is this here instead of re-using `asyncio.coroutines._is_coroutine`?

Comment (sileht) [reply Lib/inspect.py]: I moved it from asyncio to inspect as the code using it is now located in inspect module.

**Comment by sileht** [comment]
> this shouldn't be needed AsyncMock already sets
> 
> https://github.com/python/cpython/blob/5fcfdd87c9b5066a581d3ccb4b2fede938f343ec/Lib/unittest/mock.py#L2175-L2178

The test cases I added passed was passing for asyncio version, but was not for the inspect version. So something doesn't work as expected.

**Comment by graingert** [comment]
looks like this issue is because `mock.AsyncMock()` doesn't pass `inspect.isfunction`:

```python
from unittest import mock
import inspect

m = mock.AsyncMock()
with mock.patch("inspect.isfunction", new=lambda x: True):
    assert inspect.iscoroutinefunction(m)
```

_Originally posted by @graingert in https://github.com/python/cpython/issues/84753#issuecomment-1161579686_

**Comment by graingert** [comment]
Could CallableMixin mess about with isinstance/issubclass to pass `isinstance(AsyncMock(), types.FunctionType)`?

File: Lib/inspect.py (hunk 1 above)
Comment (cjw296) [reply Lib/inspect.py]: `_is_coroutine` feels like a property of asyncio, not inspect. I think it should stay where it is.

**Comment by sileht** [comment]
> Could CallableMixin mess about with `__instancecheck__/__subclasscheck__/__class__` to pass `isinstance(AsyncMock(), types.FunctionType)`?

I tried but this doesn't seem to work, I suspect this fall in `isinstance` C optimisation and `__instancecheck__` and friends are never called.

**Comment by graingert** [comment]
I'm relying on inspect.iscoroutinefunction to reveal things that really really are Coroutine Functions, partials or methods and asyncio.iscoroutinefunction to reveal things that are Duck typed as a Coroutine Function

**Comment by sileht** [comment]
> I'm relying on inspect.iscoroutinefunction to reveal things that really really are Coroutine Functions, partials or methods and asyncio.iscoroutinefunction to reveal things that are Duck typed as a Coroutine Function


According to the doc, the only difference is `asyncio.iscoroutinefunction` also works for deprecated generator based coroutine
This legacy coroutine will(have ?) be removed for 3.11, so now both functions should be the same.

**Comment by sileht** [comment]
I found another solution, I think it's cleaner.

File: Lib/test/test_inspect.py [hunk 2]
```diff
@@ -202,6 +202,10 @@ def test_iscoroutine(self):
                     gen_coroutine_function_example))))
         self.assertTrue(inspect.isgenerator(gen_coro))
 
+        self.assertFalse(
+            inspect.iscoroutinefunction(unittest.mock.Mock()))
+        self.assertTrue(
+            inspect.iscoroutinefunction(unittest.mock.AsyncMock()))
```
Comment (graingert) [inline Lib/test/test_inspect.py:208]: Probably worth adding tests for the other has_code_flags inspect tools eg isgeneratorfunction

Comment (sileht) [reply Lib/test/test_inspect.py:208]: I added one

Comment (graingert) [reply Lib/test/test_inspect.py:208]: And isasyncgenfunction?

Comment (sileht) [reply Lib/test/test_inspect.py:208]: done.

File: Lib/inspect.py [hunk 3]
```diff
@@ -395,9 +395,9 @@ def _has_code_flag(f, flag):
     while ismethod(f):
         f = f.__func__
     f = functools._unwrap_partial(f)
-    if not isfunction(f):
-        return False
-    return bool(f.__code__.co_flags & flag)
+    if isfunction(f) or _signature_is_functionlike(f):
+        return bool(f.__code__.co_flags & flag)
```
Comment (vstinner) [inline Lib/inspect.py]: Can you try to preserve current coding style? Return early with False if a condition is false? Pseudo-code:

```
if not isfunction(f): return False
if not _signature_is_functionlike(f): return False
return bool(f.__code__.co_flags & flag)
```

Comment (sileht) [reply Lib/inspect.py]: done

**Review by cjw296 — APPROVED** [review]

**Comment by sileht** [comment]
Thanks all! That was fast 🚀

**Comment by tirkarthi** [comment]
This change seems to have caused a regression reported in https://github.com/python/cpython/issues/96127
